Add V4+ features: `DATE_TYPE` and `SMALLINT_AND_TINYINT_TYPES`

dalaro opened the pull request:
This would be nice-to-have (though not essential) for the Spark connector.  The connector code necessarily checks the protocol version in several places when generating queries and thinking about schema.  These checks currently look like regular integer-primitive comparisons, which isn't ideal for readability.  Descriptively-named features are more readable, when available.

Following @adutra's  helpful preliminary comment, I split these features along OSS *C issue lines:

* https://jira.apache.org/jira/browse/CASSANDRA-7523 - date type support
* https://jira.apache.org/jira/browse/CASSANDRA-8951 - combined smallint & tinyint type support

This passed the coveo formatter enforcement and local testing (`mvn -pl core test -Dtest=DefaultProtocolVersionRegistryTest`).

olim7t commented on core/src/main/java/com/datastax/oss/driver/internal/core/DefaultProtocolFeature.java:
```diff
@@ -35,5 +35,19 @@
+   *
+   * @see <a href="https://jira.apache.org/jira/browse/CASSANDRA-7523">CASSANDRA-7523</a>
+   */
+  DATE_TYPE,
```
nit: wrong javadoc (copy-paste)

dalaro replied:
Thanks, whoops.  Looks like I introduced that [when splitting `DATE_SMALLINT_TINYINT_TYPES` into two separate enum values](https://github.com/dalaro/java-driver/commit/b3fffeef6c0a2a6704b9f537f5610b7b730e92ed).

I force-pushed a fix (and also rebased onto the latest 4.x).

dalaro committed:
Add V4+ features: `DATE_TYPE` and `SMALLINT_AND_TINYINT_TYPES`

This commit adds two DefaultProtocolFeature enum elements, supported
from V4 onward:

* SMALLINT_AND_TINYINT_TYPES

* DATE_TYPE

olim7t commented:
Squashed and rebased on new 4.5.x branch: since this has very little impact, we're targeting our upcoming 4.5.1 patch release.
